ui/src/api/api.ts

```diff
@@ -49,11 +49,17 @@ export async function listPapers(offset: number = 0, size: number = 25) {
  */
 export async function getPapers(s2Ids: string[]) {
   const PAPER_REQUEST_BATCH_SIZE = 50;
-  var s2IdsBatch = [];
-  var promises = [];
+  let s2IdsBatch = [];
+  let promises = [];
+  let dedupedIds: string[] = [];
+  s2Ids.forEach(id => {
+    if (dedupedIds.indexOf(id) < 0) {
+      dedupedIds.push(id);
+    }
+  });
 
-  for (let i = 0; i < s2Ids.length; i += PAPER_REQUEST_BATCH_SIZE) {
-    s2IdsBatch.push(s2Ids.slice(i, i + PAPER_REQUEST_BATCH_SIZE));
+  for (let i = 0; i < dedupedIds.length; i += PAPER_REQUEST_BATCH_SIZE) {
```

Comment:
I have a few recommendations on the code, but I'll let you decide what parts you want to steal.
```
async function getPapers(s2Ids) {
  // Use a set to remove dupes
  const dedupeIds = [...new Set(s2Ids)];
  
  // Use .reduce() to break into chunks
  const s2IdsBatch = dedupeIds.reduce((chunks, s2Id, index) => {
    const chunkIndex = Math.floor(index / PAPER_REQUEST_BATCH_SIZE);
    if (chunks[chunkIndex]) {
      chunks[chunkIndex].push(s2Id);
    } else {
      chunks[chunkIndex] = [s2Id];
    }
    return chunks;
  }, []);
  
  // Await the results within the async function, use async function in map
  const results = await Promise.all(s2IdsBatch.map(async (s2Ids) => {
    const result = await doGet(s2Ids);
    // ...process result
    return result;
  }));
  
  // Use `[].concat(...) to flatten array
  return [].concat(...results);
}
```

Reply:
I was operating on some outdated info that Set still had some browser incompatibilities (it still does, but with IE10 and who cares about that?). Yeah, a Set would be way better than doing what I'm doing here, and your suggestion is way cleaner.